### PR TITLE
cleanup(libsinsp): introduce set operations utils functions

### PR DESCRIPTION
**Pull request by incertum**
Signed-off-by: Melissa Kilby <melissa.kilby.oss@gmail.com>

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

> /area libscap

> /area libpman

/area libsinsp

> /area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

Introduce generic set operations functions in libs utils as discussed in https://github.com/falcosecurity/libs/pull/867, please review w/ https://github.com/falcosecurity/falco/pull/2361 use cases in mind.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: userspace/libsinsp/utils.h [hunk 1]
```diff
@@ -434,3 +436,26 @@ inline void hash_combine(std::size_t &seed, const T& val)
...
 ///////////////////////////////////////////////////////////////////////////////
 void sinsp_scap_debug_log_fn(const char* msg);
+
+///////////////////////////////////////////////////////////////////////////////
+// Set operation functions.
+///////////////////////////////////////////////////////////////////////////////
+
+
+template<typename T>
+std::set<T> unordered_set_to_ordered(std::unordered_set<T> unordered_set);
```
Comment (jasondellaluce) [inline userspace/libsinsp/utils.h:446]:
```suggestion
std::set<T> unordered_set_to_ordered(const std::unordered_set<T>& unordered_set);
```

File: userspace/libsinsp/utils.h [hunk 2]
```diff
@@ -434,3 +436,26 @@ inline void hash_combine(std::size_t &seed, const T& val)
...
+
+template<typename T>
+std::unordered_set<T> convert_vector_to_unordered_set(std::vector<T> v);
```
Comment (jasondellaluce) [inline userspace/libsinsp/utils.h:449]:
```suggestion
std::unordered_set<T> vector_to_unordered_set(const std::vector<T>& v);
```
I'm suggesting to remove the `convert_` prefix just to be coherent with the name of the other helpers. Not strong on this aspect, the important part is the const reference!

File: userspace/libsinsp/utils.h [hunk 3]
```diff
@@ -434,3 +436,26 @@ inline void hash_combine(std::size_t &seed, const T& val)
...
+
+template<typename T>
+std::unordered_set<T> unordered_set_difference(std::unordered_set<T> a, std::unordered_set<T> b);
```
Comment (jasondellaluce) [inline userspace/libsinsp/utils.h:452]:
```suggestion
std::unordered_set<T> unordered_set_difference(const std::unordered_set<T>& a, const std::unordered_set<T>& b);
```

File: userspace/libsinsp/utils.h [hunk 4]
```diff
@@ -434,3 +436,26 @@ inline void hash_combine(std::size_t &seed, const T& val)
...
+
+template<typename T>
+std::unordered_set<T> unordered_set_union(std::unordered_set<T> a, std::unordered_set<T> b);
```
Comment (jasondellaluce) [inline userspace/libsinsp/utils.h:455]:
```suggestion
std::unordered_set<T> unordered_set_union(const std::unordered_set<T>& a, const std::unordered_set<T>& b);
```

File: userspace/libsinsp/utils.h [hunk 5]
```diff
@@ -434,3 +436,26 @@ inline void hash_combine(std::size_t &seed, const T& val)
...
+
+template<typename T>
+std::unordered_set<T> unordered_set_intersection(std::unordered_set<T> a, std::unordered_set<T> b);
```
Comment (jasondellaluce) [inline userspace/libsinsp/utils.h:458]:
```suggestion
std::unordered_set<T> unordered_set_intersection(const std::unordered_set<T>& a, const std::unordered_set<T>& b);
```

File: userspace/libsinsp/utils.cpp [hunk 6]
```diff
@@ -1904,3 +1904,101 @@ void sinsp_scap_debug_log_fn(const char* msg)
...
+
+// unordered_set_difference, equivalent to SQL left_anti join operation
+template<typename T>
+std::unordered_set<T> unordered_set_difference(std::unordered_set<T> a, std::unordered_set<T> b)
+{
+	std::set<T> c = unordered_set_to_ordered(a);
+	std::set<T> d = unordered_set_to_ordered(b);
+	std::vector<T> v;
+	std::set_difference(c.begin(), c.end(), d.begin(), d.end(), std::inserter(v, v.begin()));
+	return convert_vector_to_unordered_set(v);
```
Comment (jasondellaluce) [inline userspace/libsinsp/utils.cpp:1948]:
```suggestion
	std::unordered_set<T> s;
	for(const auto& val : a)
	{
		if (b.find(val) == b.end())
		{
			s.insert(val);
		}
	}
	return s;
```
I think this is a simpler and generally more performant implementation. We don't know where these functions will be used in the future, but considering all the performance-critical paths we have I think that an extra care on performance is always good.

File: userspace/libsinsp/utils.cpp [hunk 7]
```diff
@@ -1904,3 +1904,101 @@ void sinsp_scap_debug_log_fn(const char* msg)
...
+
+// unordered_set_union
+template<typename T>
+std::unordered_set<T> unordered_set_union(std::unordered_set<T> a, std::unordered_set<T> b)
+{
+	std::set<T> c = unordered_set_to_ordered(a);
+	std::set<T> d = unordered_set_to_ordered(b);
+	std::vector<T> v;
+	std::set_union(c.begin(), c.end(), d.begin(), d.end(), std::inserter(v, v.begin()));
+	return convert_vector_to_unordered_set(v);
```
Comment (jasondellaluce) [inline userspace/libsinsp/utils.cpp:1961]:
```suggestion
	std::unordered_set<T> s = a;
	for(const auto& val : b)
	{
		s.insert(val);
	}
	return s;
```

File: userspace/libsinsp/utils.cpp [hunk 8]
```diff
@@ -1904,3 +1904,101 @@ void sinsp_scap_debug_log_fn(const char* msg)
...
+
+// unordered_set_intersection
+template<typename T>
+std::unordered_set<T> unordered_set_intersection(std::unordered_set<T> a, std::unordered_set<T> b)
+{
+	std::set<T> c = unordered_set_to_ordered(a);
+	std::set<T> d = unordered_set_to_ordered(b);
+	std::vector<T> v;
+	std::set_intersection(c.begin(), c.end(), d.begin(), d.end(), std::inserter(v, v.begin()));
+	return convert_vector_to_unordered_set(v);
```
Comment (jasondellaluce) [inline userspace/libsinsp/utils.cpp:1974]:
```suggestion
	std::unordered_set<T> s;
	for(const auto& val : a)
	{
		if (b.find(val) != b.end())
		{
			s.insert(val);
		}
	}
	return s;
```

**Review by jasondellaluce — COMMENTED** [review]
All good Melissa! My suggestions only involve performance improvements such as loop/traversal simplifications and adopting const refs consistently.

File: userspace/libsinsp/utils.cpp [hunk 9]
```diff
@@ -1920,61 +1920,104 @@ std::set<T> unordered_set_to_ordered(std::unordered_set<T> unordered_set)
...
 template<typename T>
-std::unordered_set<T> unordered_set_difference(std::unordered_set<T> a, std::unordered_set<T> b)
+std::set<T> set_difference(const std::set<T>& a, const std::set<T>& b)
 {
-	std::set<T> c = unordered_set_to_ordered(a);
-	std::set<T> d = unordered_set_to_ordered(b);
-	std::vector<T> v;
-	std::set_difference(c.begin(), c.end(), d.begin(), d.end(), std::inserter(v, v.begin()));
-	return convert_vector_to_unordered_set(v);
+	std::set<T> s;
```
Comment (jasondellaluce) [inline userspace/libsinsp/utils.cpp:1947]: For ordered `set`, I think we could rely on https://cplusplus.com/reference/algorithm/set_difference/ right away, which should also be more efficient.

File: userspace/libsinsp/utils.cpp [hunk 10]
```diff
@@ -1920,61 +1920,104 @@ std::set<T> unordered_set_to_ordered(std::unordered_set<T> unordered_set)
...
+		s.insert(val);
+	}
+	return s;
+}
+template std::unordered_set<std::string> unordered_set_union(const std::unordered_set<std::string>& a, const std::unordered_set<std::string>& b);
+template std::unordered_set<uint32_t> unordered_set_union(const std::unordered_set<uint32_t>& a, const std::unordered_set<uint32_t>& b);
+
+// set_union
+template<typename T>
+std::set<T> set_union(const std::set<T>& a, const std::set<T>& b)
```
Comment (jasondellaluce) [inline userspace/libsinsp/utils.cpp:1976]: Same here for https://cplusplus.com/reference/algorithm/set_union/.

File: userspace/libsinsp/utils.cpp [hunk 11]
```diff
@@ -1920,61 +1920,104 @@ std::set<T> unordered_set_to_ordered(std::unordered_set<T> unordered_set)
...
+		}
+	}
+	return s;
+}
+template std::unordered_set<std::string> unordered_set_intersection(const std::unordered_set<std::string>& a, const std::unordered_set<std::string>& b);
+template std::unordered_set<uint32_t> unordered_set_intersection(const std::unordered_set<uint32_t>& a, const std::unordered_set<uint32_t>& b);
+
+// set_intersection
+template<typename T>
+std::set<T> set_intersection(const std::set<T>& a, const std::set<T>& b)
```
Comment (jasondellaluce) [inline userspace/libsinsp/utils.cpp:2007]: Same here for https://cplusplus.com/reference/algorithm/set_intersection/.

**Review by jasondellaluce — COMMENTED** [review]
Final remarks, just suggesting using the STL implementation for `set`, which is more performant for ordered iterables. Besides that, I think this is good and ready for mainline.

**Review by jasondellaluce — APPROVED** [review]
/approve

**Review by FedeDP — APPROVED** [review]
/approve

**Pull request merged**